### ISSUE TITLE
A design idea

### ISSUE THREAD

**Issue by tiotrom** [SimonBrazell/privacy-redirect#98]
What if, in the drop down menu, you not only add the disable/enable instances, but also the ability to select them. The drop down menu for selecting the instances. Either leave it as random, or select a specific one. Since at times some instances do not work, you are forced to go to the "more options" and change the instance. Could be super useful to have these options quickly accessible. Cheers!

**Comment by SimonBrazell** [comment]
Hi @tiotrom this is how it used to be, but someone suggested I remove the instances to reduce clutter in the pop-up menu, putting everything instead in the options.

**Comment by tiotrom** [comment]
I understand the reasoning, but it could be very useful to have them all in one drop down menu. However if you add the "random" pull for instances, it may not be as necessary.

**Comment by SimonBrazell** [comment]
The random instance change is made, look out for v1.1.41.

**Comment by Cade66** [comment]
I'm in favor of adding back the instances to the pop-up menu, even if I understand the developer's motivation.
Invidious is having a hard time, its main developer has ended his essential participation, instances regularly happen to encounter problems when others recover, so to say; this means the user is often obliged to switch from one instance to the other. Privacy Redirect's Options called from the pop-up links to it's page on Firefox's Password Manager and from there the user has to click again on Options (or create a direct link in the Bookmarks Toolbar to open directly those Options, as i did). A bit cumbersome compared to direct Instance selection right from the toolbar button's pop-up, as I see it.

**Comment by SimonBrazell** [comment]
Yeah I'll add them back, wish I didn't remove it in the first place.

> Privacy Redirect's Options called from the pop-up links to it's page on Firefox's Password Manager and from there the user has to click again on Options

This sounds a little odd to me, certainly haven't seen this in the tests I've run locally. Which version are you running? This sounds like a Firefox bug tbh, the link on the popup should go directly to the preferences / options for the extension.

**Comment by Cade66** [comment]
`This sounds a little odd to me, certainly haven't seen this in the tests I've run locally. Which version are you running? This sounds like a Firefox bug tbh, the link on the popup should go directly to the preferences / options for the extension.`

I'm running latest Privacy Redirect 1.1.41 / Firefox 80.0.1 / Win7.
But this is not specific to Privacy Redirect, it concerns other extensions as well, whilst not all, and I never understood by the way why some and not others.
For instance the LocalCDN Firefox extension behaves as Privacy Redirect : clicking on on its toolbar button deploys a drop-down menu which includes an 'Options' button which doesn't open directly the Options page but the extension's page in the Password Manager. I have over 40 extensions and for those concerned ('options' availability), most open directly their own options page (something like `moz-extension://[Internal UUID]/[...]options.html`) and others not... never understood why, not being a programmer myself.

Of course if PR' toolbar-button's drop-down menu's 'More Options' opened directly its More Options page (`moz-extension://[Internal UUID]/pages/options/options.html`) then  having Invidious instances right in the pop-up menu would be less pertinent and I imagine this logic participated to the move in the first place.

All this is scondary of course. The main thing is that Privacy Redirect ... redirects correctly, and it does!

**Comment by SimonBrazell** [comment]
> I have over 40 extensions and for those concerned ('options' availability), most open directly their own options page (something like moz-extension://[Internal UUID]/[...]options.html) and others not... never understood why, not being a programmer myself.

@Cade66 I think I understand why, it's a decision that the developer can make to either load the options within the browser's inbuilt extension options UI (which is what Privacy Redirect does) or open their own in a new tab / window. 

My reasoning for using the inbuilt UI was that my options page isn't that complex and so could be handled by that instead of doing it in a new tab, but I could just do it the other way, not a huge change. Perhaps this issue is why the majority of the others do it that way instead?

**Comment by SimonBrazell** [comment]
Curiously though, when I click the "More Options" button on the pop-up using Firefox on my work machine (macOS), I go straight to the following page as expected:

![Screen Shot 2020-09-09 at 3 26 07 pm](https://user-images.githubusercontent.com/16571951/92557954-d29ea680-f2b0-11ea-8e2f-fbbd7d7961c7.png)

**Comment by SimonBrazell** [comment]
So if I just fix whatever this underlying issue is with the "More Options" button, possibly by loading the preferences in a new tab for Firefox only (are you able to confirm this doesn't happen on Chromium also for your machine?), then it would be fine to leave the instance inputs out of the pop-up?

**Comment by SimonBrazell** [comment]
It's as simple as changing this in the `manifest.json` file I think:

```
"options_ui": {
    "page": "pages/options/options.html",
    "open_in_tab": false <------ change to true
  },
```

**Comment by SimonBrazell** [comment]
And then it opens in a new tab looking something like this instead:

![Screen Shot 2020-09-09 at 3 38 03 pm](https://user-images.githubusercontent.com/16571951/92558730-80f71b80-f2b2-11ea-9d0f-406d8f176fa4.png)

**Comment by Cade66** [comment]
`I think I understand why, it's a decision that the developer can make to either load the options within the browser's inbuilt extension options UI (which is what Privacy Redirect does) or open their own in a new tab / window.`
Acknowledged! That's a point I discover. I remain puzzled that it doesn't explain why the problem doesn't appear on macOS as you've noted:

`Curiously though, when I click the "More Options" button on the pop-up using Firefox on my work machine (macOS), I go straight to the following page as expected:`
Odd indeed. This is with Firefox on macOS whilst my experience is performed on Windows (7), which maybe (likely) explains it all?

`So if I just fix whatever this underlying issue is with the "More Options" button, possibly by loading the preferences in a new tab for Firefox only (are you able to confirm this doesn't happen on Chromium also for your machine?), then it would be fine to leave the instance inputs out of the pop-up?`
I cannot confirm this doesn't happen on Chromium on my machine given Chromium is not installed here.
My humble opinion is that it would be fine indeed to leave the instance inputs out of the pop-up given the Options link would open directly the Options' input and not the extension's page in the Add-on Manager. This would satisfy everyone. The idea is to access the input as directly as possible.

Thanks again!

**Comment by tiotrom** [comment]
I can add that on Linux and Firefox, it opens directly the Options page for the extension. So it works perfectly fine for me.

**Comment by Cade66** [comment]
@tiotrom wrote, `I can add that on Linux and Firefox, it opens directly the Options page for the extension. So it works perfectly fine for me.`
@SimonBrazell (the developer) wrote, `Curiously though, when I click the "More Options" button on the pop-up using Firefox on my work machine (macOS), I go straight to the following page as expected`

I'm heading to a magic mystery conclusion that the "_issue_" is related to the _Firefox & Windows_ partnership...

**Comment by SimonBrazell** [comment]
> I'm heading to a magic mystery conclusion that the "issue" is related to the Firefox & Windows partnership...

I'd say that the issue is more accurately Windows 7 + Firefox as I have confirmed the correct behaviour on Windows 10 Firefox and Chromium.

> Yet, he points out that modifying the manifest.json file resolves the issue on Firefox/Windows ...

The problem with modifying the manifest file is that it applies to all environments, I thought that perhaps there was a way to do it just for Firefox but doesn't seem to be the case, all good though it's not a huge change, can add it to the next build and see if it helps, if not I'll revert and then look at adding the instances back to the pop-up, how's that sound?

**Comment by Cade66** [comment]
> I'd say that the issue is more accurately Windows 7 + Firefox as I have confirmed the correct behaviour on Windows 10 Firefox and Chromium.

The equation is reducing; that's how a fellow resolved energy to mc^2 lol.

Yet I linger to understand why, then, Firefox extensions' `manifest.json` includes that `"options_ui"` mentioned above. What's the point if it only interferes/concerns Win7?
Maybe just not changing anything to the extension, given the non-issue of what appears to be a slight bother on Win7 only, is the best approach; I don't know.

I'm sure you'll do for the best. Meanwhile this thread will at least have emphasized on this odd extension to its options deployment via or not the Add-on Manager, concerning some extensions and not others and apparently only when performed on Win7. I remain puzzled.

**Comment by SimonBrazell** [comment]
> Yet I linger to understand why, then, Firefox extensions' manifest.json includes that "options_ui" mentioned above. 

The option is included as a global setting affecting all browsers and OS, it allows you to open the extension preferences in a new tab instead of the extension preferences / options page within the browser extensions UI. It's not just a fix for Win7 although I suspect it would help with your problem.

I'm not sure what to do here, my preference is to not change this or the pop-up menu, but I probably don't use Invidious etc. as much as some others like @tiotrom lately, my experience with the random instances has been decent so far with the little use that it sees.

**Comment by Cade66** [comment]
`The option is included as a global setting affecting all browsers and OS, it allows you to open the extension preferences in a new tab instead of the extension preferences / options page within the browser extensions UI. It's not just a fix for Win7 although I suspect it would help with your problem.`

I think that's quite clear, yet why this option if it were to concern only Windows 7? That's what I don't understand.

`I'm not sure what to do here [...]`

IMHO? Nothing. The extension works fine, no need to get on a brain storm for what is finally a peccadillo. In my case I have a set a bookmark in the Bookmarks Toolbar which points directly to Privacy Redirect's Options. No fuss.

As far as i'm concerned this problematic, though interesting, is closed. Thanks for your commitment as well of course for the extension, best in its category.
